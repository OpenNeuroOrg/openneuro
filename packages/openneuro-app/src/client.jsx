/**
 * Browser client entrypoint - see server.tsx for SSR entrypoint
 */
import './scripts/utils/global-polyfill'
import { setupApm } from './scripts/apm.js'
import * as Sentry from '@sentry/browser'
import { ApolloProvider, InMemoryCache } from '@apollo/client'
import { createClient } from '@openneuro/client'
import React from 'react'
import ReactDOM from 'react-dom'
import { BrowserRouter, Route } from 'react-router-dom'
import App from './scripts/app'
import Index from './scripts/index'
import analyticsWrapper from './scripts/utils/analytics'
import { version } from './lerna.json'
import { config } from './scripts/config'
import * as gtag from './scripts/utils/gtag'

if (
  config.sentry.environment === 'production' ||
  config.sentry.environment === 'staging'
) {
  setupApm({
    serverUrl: config.url,
    serviceName: '@openneuro/app',
    serviceVersion: version,
    environment: config.sentry.environment,
  })
} else {
  setupApm({
    serverUrl: config.url,
    serviceName: '@openneuro/app',
    serviceVersion: version,
    environment: config.sentry.environment,
    active: false,
  })
}

gtag.initialize(config.analytics.trackingIds)

Sentry.init({
  dsn: 'https://ba0c58863b3e40a2a412132bfd2711ea@sentry.io/251076',
  release: version,
  environment: config.sentry.environment,
})

// Setup the service worker
if ('serviceWorker' in navigator) {
  //navigator.serviceWorker.register('/sw.js')
} else {
  Sentry.captureMessage('Service worker registration failed.')
}

ReactDOM.render(
  <App>
    <ApolloProvider
      client={createClient(`${config.url}/crn/graphql`, {
        clientVersion: version,
        // @ts-expect-error
        cache: new InMemoryCache().restore(JSON.parse(window.__APOLLO_STATE__)),
      })}>
      <BrowserRouter>
        <Route component={analyticsWrapper(Index)} />
      </BrowserRouter>
    </ApolloProvider>
  </App>,
  document.getElementById('main'),
)
