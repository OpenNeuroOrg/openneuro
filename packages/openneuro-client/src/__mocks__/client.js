import { ApolloClient, InMemoryCache } from '@apollo/client'
import { SchemaLink } from '@apollo/client/link/schema'
import { addMockFunctionsToSchema, makeExecutableSchema } from 'apollo-server'
import { typeDefs } from '@openneuro/server/src/graphql/schema'
import resolvers from '@openneuro/server/src/graphql/resolvers/index.js'

/**
 * Generate sequential dataset ids for tests
 * @param {number} n Starting string value
 */
function* idGenerator(n = 0) {
  while (true) {
    const idStr = String(++n).padStart(6, '0')
    yield `ds${idStr}`
  }
}

export const datasetIds = idGenerator()
export const testDsId = datasetIds.next().value
export const testTime = new Date()

// Workaround to restore this test, it should be refactored with openneuro-client updates
const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
})

addMockFunctionsToSchema({
  schema,
  mocks: {
    Dataset: (root, { id }) => ({
      id: () => (id ? id : datasetIds.next().value),
      created: testTime,
      modified: testTime,
    }),
    DateTime: () => {
      return testTime
    },
    BigInt: () => {
      return 8589934592
    },
  },
})

export const createClient = uri => {
  const cache = new InMemoryCache()
  const link = new SchemaLink({
    schema,
  })
  // @ts-expect-error
  return new ApolloClient({ uri, link, cache })
}
