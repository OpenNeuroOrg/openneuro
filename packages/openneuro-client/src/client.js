import fetch from 'node-fetch'
import ApolloClient from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { createUploadLink } from 'apollo-upload-client/src'
import FormData from 'form-data'
import * as files from './files'
import * as datasets from './datasets'

const cache = new InMemoryCache()

/**
 * Setup a client for working with the OpenNeuro API
 *
 * @param {string} uri
 */
const createClient = uri => {
  const link = createLink(uri)
  return new ApolloClient({ uri, link, cache })
}

const createLink = uri => {
  return createUploadLink({ uri, fetch, serverFormData: FormData })
}

export { files, datasets }
export default createClient
