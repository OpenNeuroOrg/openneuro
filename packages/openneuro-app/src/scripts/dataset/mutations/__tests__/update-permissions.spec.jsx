import React from "react"
import { fireEvent, render, screen, waitFor } from "@testing-library/react"
import { MockedProvider } from "@apollo/client/testing"
import {
  UPDATE_ORCID_PERMISSIONS,
  UPDATE_PERMISSIONS,
  UpdateDatasetPermissions,
} from "../update-permissions"

import { isValidOrcid } from "../../../utils/validationUtils.ts";

function permissionMocksFactory(
  updatePermissionsCalled,
  updateOrcidPermissionsCalled,
) {
  return [
    {
      request: {
        query: UPDATE_PERMISSIONS,

        variables: {
          datasetId: "ds000005",
          userEmail: "test@example.com",
          level: "ro",
        },
      },
      newData: updatePermissionsCalled,
    },
    {
      request: {
        query: UPDATE_ORCID_PERMISSIONS,

        variables: {
          datasetId: "ds000005",
          userOrcid: "0000-0002-1694-233X",
          level: "ro",
        },
      },
      newData: updateOrcidPermissionsCalled,
    },
  ]
}

describe("UpdateDatasetPermissions mutation", () => {
  it("renders with default props", () => {
    const { asFragment } = render(
      <MockedProvider>
        <UpdateDatasetPermissions />
      </MockedProvider>,
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("calls UPDATE_PERMISSIONS when clicked with an email address", async () => {
    const updatePermissionsCalled = vi.fn()
    const updateOrcidPermissionsCalled = vi.fn()
    const mocks = permissionMocksFactory(
      updatePermissionsCalled,
      updateOrcidPermissionsCalled,
    )
    const done = vi.fn()
    const { asFragment } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <UpdateDatasetPermissions
          datasetId="ds000005"
          userIdentifier="test@example.com"
          metadata="ro"
          done={done}
        />
      </MockedProvider>,
    )
    const fragment = asFragment()
    expect(fragment).toMatchSnapshot()
    // Try clicking the button and make sure the right mutation runs
    const button = screen.getByRole("button")
    await fireEvent.click(button)
    // Make sure it ran at all
    await waitFor(() => expect(done).toHaveBeenCalled())
    // Verify the expected query ran
    await waitFor(() => expect(updatePermissionsCalled).toHaveBeenCalled())
    await waitFor(() =>
      expect(updateOrcidPermissionsCalled).not.toHaveBeenCalled()
    )
  })
  it("calls UPDATE_ORCID_PERMISSIONS when clicked with an ORCID", async () => {
    const updatePermissionsCalled = vi.fn()
    const updateOrcidPermissionsCalled = vi.fn()
    const mocks = permissionMocksFactory(
      updatePermissionsCalled,
      updateOrcidPermissionsCalled,
    )
    const done = vi.fn()
    const { asFragment } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <UpdateDatasetPermissions
          datasetId="ds000005"
          userIdentifier="0000-0002-1694-233X"
          metadata="ro"
          done={done}
        />
      </MockedProvider>,
    )
    const fragment = asFragment()
    expect(fragment).toMatchSnapshot()
    // Try clicking the button and make sure the right mutation runs
    const button = screen.getByRole("button")
    await fireEvent.click(button)
    // Make sure it ran at all
    await waitFor(() => expect(done).toHaveBeenCalled())
    // Verify the expected query ran
    await waitFor(() => expect(updatePermissionsCalled).not.toHaveBeenCalled())
    await waitFor(() => expect(updateOrcidPermissionsCalled).toHaveBeenCalled())
  })
  describe("isValidOrcid", () => {
    it("matches typical ORCID strings", () => {
      expect(isValidOrcid("0000-0001-2345-678")).toBe(false)
      expect(isValidOrcid("0000-0001-2345-678f")).toBe(false)
      expect(isValidOrcid("19818c4d-1e60-4480-a317-6fcc1c1a88c6")).toBe(false)
      expect(isValidOrcid("0000000123456789")).toBe(false)
      // Check for a correct value
      expect(isValidOrcid("0000-0001-2345-6789")).toBe(true)
      // Test with the X checksum value
      expect(isValidOrcid("0000-0002-1694-233X")).toBe(true)
    })
  })
})
