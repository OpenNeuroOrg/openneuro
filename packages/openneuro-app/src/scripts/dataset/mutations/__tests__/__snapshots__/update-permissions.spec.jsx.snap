// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UpdateDatasetPermissions mutation > calls UPDATE_ORCID_PERMISSIONS when clicked with an ORCID 1`] = `
<DocumentFragment>
  <button
    aria-label="Share"
    class="on-button on-button--small on-button--primary  btn-modal-action"
    role="button"
    type="button"
  >
    Share
  </button>
</DocumentFragment>
`;

exports[`UpdateDatasetPermissions mutation > calls UPDATE_PERMISSIONS when clicked with an email address 1`] = `
<DocumentFragment>
  <button
    aria-label="Share"
    class="on-button on-button--small on-button--primary  btn-modal-action"
    role="button"
    type="button"
  >
    Share
  </button>
</DocumentFragment>
`;

exports[`UpdateDatasetPermissions mutation > renders with default props 1`] = `
<DocumentFragment>
  <button
    aria-label="Share"
    class="on-button on-button--small on-button--primary  btn-modal-action"
    role="button"
    type="button"
  >
    Share
  </button>
</DocumentFragment>
`;
